### DISCUSSION TITLE
CORS-friendly URLs now available via GitHub Pages

### DISCUSSION THREAD

**Discussion by matteason** [matteason/daily-cloud-maps#1] [Announcements]
Previously, the daily images published by this project couldn't be used directly in JavaScript because GitHub release assets don't include an `Access-Control-Allow-Origin: *` header.

I'd investigated publishing the images to GitHub pages before (which does set the necessary header), but the only publishing method available stored the published page content in a branch - meaning every image published would become part of the repository's history. With each day's images totalling around 80MB, this would have meant adding around 1GB of history to the repo every 12 days.

Luckily, GitHub [now allow us to push arbitrary static content to Pages](https://github.blog/changelog/2022-07-27-github-pages-custom-github-actions-workflows-beta/) without it being in the repo, meaning no ballooning history. So each day's images are now published to GitHub pages, replacing the previous day's images.

I've updated the [readme](https://github.com/matteason/daily-cloud-maps) with the new URLs.

**Comment by mikehardy** [comment]
This has nothing to do with your post, and I'm sorry for that, but I just wanted to say thanks! I just integrated this (simple wget via cron, then xplanet) and I have up to date clouds on my earth background. Much appreciated!

One thing I've noticed is that the brightness on the day map and the darkness on the dark map seem really exaggerated, but I think that's probably something I'm doing. How are you using these, if I might ask?

I wonder if I should just do some post-processing to tune them the way I like or if this is something nagging at you or not?

**Reply by matteason** [reply]
Hi @mikehardy, thanks for using it! On the brightness - I'm using the `earth` and `earth-night` images as textures in a three.js app. I'm using a custom shader which does the night/day mixing and doesn't use any other lighting sources, so I wonder if xplanet is simulating the sun with a light source which is blowing out the highlights on the day map and causing the night map to be completely unlit? Do you mind sharing a screenshot?

**Reply by mikehardy** [reply]
sure - here's my details. I'm not sitting here with a secret expectation that you'll make my use case perfect, but if you ask and offer I'll happily take it. Cloud nerds of the world unite ;-)

![rectangular_1661352601](https://user-images.githubusercontent.com/782704/186451312-b31c78d6-8992-4ac5-a7e0-495b01466c3f.jpg)

Pretty boring setup I think, this may be useful as "how to use it" docs for others though, this sets background on an ubuntu machine (maybe any gnome desktop?) via cron execution

```bash
#!/bin/bash

xplanetcmd="/usr/bin/xplanet"
config="/home/mike/.xplanet/xplanet_config"
outdir="/home/mike/Pictures/xplanet"
date=`date +%s`

# Check if the output dir exists
if [ ! -d "$outdir" ]; then
	mkdir -p "$outdir"
fi

# Create the basic output (change the resolution to whatever width
# your screen is and half that for the height assuming you have
# source images with an aspect ratio of 2:1 of course...
$xplanetcmd -searchdir ~/.xplanet -config $config \
--quality 95 -verbosity 0 --num_times 1 \
--output $outdir/rectangular_$date.jpg --geometry 1920x960 \
--projection rectangular

# Now set the background
USERID=`id -u`
export DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/$USERID/bus

IMAGE_NAME=`ls -1 -tr $outdir | tail -n 1`
IMAGE_FILE=$outdir/$IMAGE_NAME
gsettings set org.gnome.desktop.background picture-uri "file://$IMAGE_FILE"
gsettings set org.gnome.desktop.background picture-uri-dark "file://$IMAGE_FILE"
gsettings set org.gnome.desktop.screensaver picture-uri "file://$IMAGE_FILE"

# Finally clean up old files
pushd $outdir
if [ `ls -1 | wc -l` != 1 ]; then
  ls -t $outdir | tail -n +2 | xargs rm -- 
fi
popd

```

`xplanet_config`:

```
[earth]
map=4096x2048-earth.jpg
night_map=4096x2048-earth-night.jpg
```

**Reply by matteason** [reply]
Oh interesting, ok - that look pretty much how I'd expect. The day brightness is the same as the original Blue Marble image. For the night image I reverse the cloud mask so that the clouds are black before overlaying it. That makes the lights obscured by cloud, but it'll make most of it darker than the night image without clouds. You could set your night image to the unmodified one (https://www.nasa.gov/specials/blackmarble/2012/globalmaps/BlackMarble_2012_01deg.jpg) if you'd like brighter lights or combine the PNG cloud map and the night map with something like imagemagick if you want the clouds to be a bit more subtle

**Reply by mikehardy** [reply]
Thanks for the pointer - I'll look into a little post-processing to make it subtle then since this matches your expectation.
It doesn't look incorrect to me either really, just a bit jarring as a desktop background, but that's my preference

Thanks again for setting this up! Cheers

**Reply by matteason** [reply]
No problem! You might want to play with the `twilight` config option to make the day/night transition a bit less harsh (assuming it works with rectangular projection):

>twilight
>Let the day and night hemispheres blend into one another for pixels
>within the specified number of degrees of the terminator.  The default
>value is 6.

**Reply by maitra** [reply]
> This has nothing to do with your post, and I'm sorry for that, but I just wanted to say thanks! I just integrated this (simple wget via cron, then xplanet) and I have up to date clouds on my earth background. Much appreciated!
> 
> One thing I've noticed is that the brightness on the day map and the darkness on the dark map seem really exaggerated, but I think that's probably something I'm doing. How are you using these, if I might ask?
> 
> I wonder if I should just do some post-processing to tune them the way I like or if this is something nagging at you or not?

Do you mind sharing your wget script? I more often than not (but not every day) get file fully retrieved, nothing to do so I am clearly not doing something right in my calls. Thanks!

**Reply by wlbr** [reply]
Curl instead of wget: `curl https://matteason.github.io/daily-cloud-maps/2048x1024-clouds.jpg -o ~/Library/Xplanet/clouds2k.jpg`
I scheduled the download to 15:45UTC, so that I can be more or less sure, that there are new files at the origin.

**Reply by matteason** [reply]
Thanks for using the project @maitra and @wlbr !

> I more often than not (but not every day) get file fully retrieved

Are you using the permalink to the latest image (as in @wlbr's comment) or one of the URLs with a date in (for example `https://github.com/matteason/daily-cloud-maps/releases/download/2023-02-19/8192x4096-clouds.jpg`)?

Bear in mind that occasionally there might be a problem generating an image for a particular day. This is usually when there's a long delay in the satellite data being made available in GIBS (the API this project gets its data from). In those cases the dated URL will return a 404 and the latest image permalink will still work, but just serve the last successfully generated image, which may result in wget saying you already have the latest copy of the file as it hasn't changed on the server. The timing of when a new image is available can also be variable because GitHub Actions aren't guaranteed to run on time and the image generation time can also vary.

**Reply by maitra** [reply]
> Thanks for using the project @maitra and @wlbr !
> 
> > I more often than not (but not every day) get file fully retrieved
> 
> Are you using the permalink to the latest image (as in @wlbr's comment) or one of the URLs with a date in (for example `https://github.com/matteason/daily-cloud-maps/releases/download/2023-02-19/8192x4096-clouds.jpg`)?
> 
> Bear in mind that occasionally there might be a problem generating an image for a particular day. This is usually when there's a long delay in the satellite data being made available in GIBS (the API this project gets its data from). In those cases the dated URL will return a 404 and the latest image permalink will still work, but just serve the last successfully generated image, which may result in wget saying you already have the latest copy of the file as it hasn't changed on the server. The timing of when a new image is available can also be variable because GitHub Actions aren't guaranteed to run on time and the image generation time can also vary.

Thanks, I do it  only about an hour or more after the maps are posted (I get email every day). I simply use wget -c --no-check-certificate https://matteason.github.io/daily-cloud-maps/8192x4096-clouds.jpg (btw, should I try  with a date everyday?)

```

 wget -c https://matteason.github.io/daily-cloud-maps/8192x4096-clouds.jpg
--2023-02-22 07:44:02--  https://matteason.github.io/daily-cloud-maps/8192x4096-clouds.jpg
Resolving matteason.github.io (matteason.github.io)... 2606:50c0:8002::153, 2606:50c0:8001::153, 2606:50c0:8000::153, ...
Connecting to matteason.github.io (matteason.github.io)|2606:50c0:8002::153|:443... connected.
HTTP request sent, awaiting response... 416 Range Not Satisfiable

    The file is already fully retrieved; nothing to do.

```

I will try curl from now on, and report back if there are issues.
